src/repo/repository.rs

```diff
@@ -116,6 +117,13 @@ pub struct ExtractResult {
 pub struct Repository {
     /// The path containing the [`Repository::data_dir`] directory.
     path: PathBuf,
+    /// With the addition of `elfshaker clone`, there are now longer running tasks
+    /// for which it would be useful to provide the user with a progress bar.
+    /// Since there might be multiple long running sub-tasks invoked in each
+    /// macro tasks (e.g. extract snapshot includes fetching the .esi,
+    /// fetching individual pack, etc.), it is useful to use a "factory",
+    /// instead of argument passing for the [`ProgressReporter`].
+    progress_reporter_factory: Box<dyn Fn(&str) -> ProgressReporter<'static> + Send + Sync>,
```

Comment:
I don't feel great about putting this state on the repository if it can be avoided, since it doesn't feel like repository state (application state, maybe). That said, I understand there are tradeoffs to make here so I'm not going to request you change it.

Reply:
This is not really state as much as it is configuration. It is a progress reporter that is local to the repository object and can be used internally to print progress. 

Note that the progress information and the methods we print it from might change with time. So instead of passing in a `ProgressReporter` to every public method, we set it for the whole `Repository`. The second-best alternative would be to use a global.

Reply:
I'm not proposing to change this now, though perhaps there are refinements we could do to progress reporter if t
Just to flesh out my undestanding: I can imagine a case where we might have more than one repository. Would this work as is today, or would it be broken? If it were a global, would it be broken?

When I say 'state', in my mind, I mean that the repository should only really know about things which relate to the repository. Putting something unrelated to it seems strange. The 'struct' which holds knowledge of things relating to application state might be better called `App` in my mind. I think I would object less to the `Repository` holding a reference to an `App` so that it could print things.

I don't particularly like the idea of the reporter factory being a global state, either, from simply an aversion to having global state. Howeer, it's intersting to entertain the idea -- is it really so different from `println!()` and friends, which effectively access a writer through magic? Possibly; yes it is, because `println!()` presumably doesn't hold any state other than perhaps buffers, contrasting with the progress reporter, which is keeping track of how much work has been done.

Something else I would like to see tried is if the reporter factory were simply passed-by-argument instead. Do we have a sense of how bad that would be? How many function signatures would need to be modified? If it were, say, less than 10, I might even favour simply passing it in.

Reply:
I gave you incorrect information in my comment above which confused the both of us -- sorry about that. 

> It is a progress reporter **++factory++** that is local to the repository object and can be used internally to **++create ProgressReporters which can be used to++** print progress.

This `progress_reporter_factory` doesn't hold state. It holds a boxed callable which creates a `ProgressReporter` -- it is a stateless factory. The progress reporter is what holds the state, but ProgressReporters are created locally, by calling `(self.progress_reporter_factory)(...)`. The ProgressReporters are passed as arguments. The factory is configured from the top-level `run` method:

https://github.com/elfshaker/elfshaker/blob/627f704ead489047132dbab593c8151df8a45619/src/bin/elfshaker/clone.rs#L47